src/Aeson.purs

```diff
@@ -112,24 +165,156 @@ getField' decoder obj str =
+  unsafeIndex arr ix = unsafePartial $ Array.unsafeIndex arr ix
+
+  -- will never encounter non int number
+  coerceNumber :: Number -> Int
+  coerceNumber = round
+
+constAesonCases :: forall (a :: Type). a -> AesonCases a
+constAesonCases v =
+  { caseObject: c, caseNull: c, caseBoolean: c, caseString: c, caseNumber: c, caseArray: c }
+  where
+    c :: forall (b :: Type). b -> a
+    c = const v
+
+toObject :: Aeson -> Maybe (Object Aeson)
+toObject =
+  caseAeson $ constAesonCases Nothing # _ { caseObject = Just }
+
+-------- Decode helpers --------
 
 -- | Ignore numeric index and reuse Argonaut decoder.
 decodeAesonViaJson
   :: forall (a :: Type). DecodeJson a => Aeson -> Either JsonDecodeError a
-decodeAesonViaJson (Aeson { json }) = decodeJson json
+decodeAesonViaJson (Aeson { patchedJson: AesonPatchedJson j }) = decodeJson j
 
 decodeAesonString
   :: forall (a :: Type). DecodeAeson a => String -> Either JsonDecodeError a
-decodeAesonString payload = do
-  json <- note MissingValue $ hush $ jsonParser patchedPayload
-  decodeAeson (Aeson { index, json })
-  where
-  patchedPayload /\ index = parseJsonExtractingIntegers Tuple payload
+decodeAesonString = parseJsonStringToAeson >=> decodeAeson
+
+-------- DecodeAeson instances --------
+
+instance DecodeAeson Int where
+  decodeAeson aeson@(Aeson { numberIndex }) = do
+    -- Numbers are replaced by their index in the array.
+    ix <- decodeAesonViaJson aeson
+    numberStr <- note MissingValue (numberIndex Array.!! ix)
+    note MissingValue $ Int.fromString numberStr
+
+instance DecodeAeson BigInt where
+  decodeAeson aeson@(Aeson { numberIndex }) = do
+    -- Numbers are replaced by their index in the array.
+    ix <- decodeAesonViaJson aeson
+    numberStr <- note MissingValue (numberIndex Array.!! ix)
+    note MissingValue $ BigInt.fromString numberStr
+
+instance DecodeAeson Boolean where
+  decodeAeson = decodeAesonViaJson
+
+instance DecodeAeson String where
+  decodeAeson = decodeAesonViaJson
+
+instance DecodeAeson Number where
+  decodeAeson = decodeAesonViaJson
+
+instance DecodeAeson Json where
+  decodeAeson = Right <<< toStringifiedNumbersJson
+
+instance DecodeAeson Aeson where
+  decodeAeson = pure
+
+instance (GDecodeAeson row list, RL.RowToList row list) => DecodeAeson (Record row) where
+  decodeAeson json =
+    case toObject json of
+      Just object -> gDecodeAeson object (Proxy :: Proxy list)
+      Nothing -> Left $ TypeMismatch "Object"
+
+else instance (Traversable t, DecodeAeson a, DecodeJson (t Json)) => DecodeAeson (t a) where
+  decodeAeson (Aeson { numberIndex, patchedJson: AesonPatchedJson pJson }) = do
+    jsons :: t _ <- map AesonPatchedJson <$> decodeJson pJson
+    for jsons (\patchedJson -> decodeAeson (Aeson { patchedJson, numberIndex }))
+
+class GDecodeAeson (row :: Row Type) (list :: RL.RowList Type) | list -> row where
+  gDecodeAeson :: forall proxy. FO.Object Aeson -> proxy list -> Either JsonDecodeError (Record row)
```

Comment:
Minor aesthetic but there's `FO.Object` and `Object` in the module due to the import
```
import Foreign.Object (Object)
import Foreign.Object as FO
```
The import is fine, but Is it worth sticking to one in the codebase?